### ISSUE TITLE
Support for diagnostics = TRUE

### ISSUE THREAD

**Issue by grantmcdermott** [vincentarelbundock/modelsummary#123]
Dude, sorry to be a pain with these regular issues. 

(What can I say? I love the package and you're just so efficient at fixing / adding functionality ;-) )

I don't want to push too hard &mdash; especially since what I'm about to say could conceivably be covered by the custom tidy methods &mdash; but I think there's an argument for supporting the "diagnostics" argument of `glance()` out of the box.

Ofc, a lot of models don't even support a diagnostics argumment. But for those that do, this is often very helpful information. Consider our friend from yesterday's discussion, IV regression, but this time using the ivreg package.

```r
library(ivreg)
library(broom)

demand = ivreg(Q ~ P + D | D + F + A, data = Kmenta)

## Note "Diagnostics tests" towards bottom of summary printout
summary(demand)
#> 
#> Call:
#> ivreg(formula = Q ~ P + D | D + F + A, data = Kmenta)
#> 
#> Residuals:
#>     Min      1Q  Median      3Q     Max 
#> -3.4305 -1.2432 -0.1895  1.5762  2.4920 
#> 
#> Coefficients:
#>             Estimate Std. Error t value Pr(>|t|)    
#> (Intercept) 94.63330    7.92084  11.947 1.08e-09 ***
#> P           -0.24356    0.09648  -2.524   0.0218 *  
#> D            0.31399    0.04694   6.689 3.81e-06 ***
#> 
#> Diagnostic tests:
#>                  df1 df2 statistic  p-value    
#> Weak instruments   2  16    88.025 2.32e-09 ***
#> Wu-Hausman         1  16    11.422  0.00382 ** 
#> Sargan             1  NA     2.983  0.08414 .  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Residual standard error: 1.966 on 17 degrees of freedom
#> Multiple R-Squared: 0.7548,  Adjusted R-squared: 0.726 
#> Wald test: 23.81 on 2 and 17 DF,  p-value: 1.178e-05

## To view diagnostics with glance, add diagnostics = TRUE
glance(demand, diagnostics = TRUE )
#> # A tibble: 1 x 12
#>   r.squared adj.r.squared sigma statistic p.value    df df.residual  nobs
#>       <dbl>         <dbl> <dbl>     <dbl>   <dbl> <int>       <int> <int>
#> 1     0.755         0.726  1.97      23.8 1.18e-5     3          17    20
#> # … with 4 more variables: statistic.Sargan <dbl>, p.value.Sargan <dbl>,
#> #   statistic.Wu.Hausman <dbl>, p.value.Wu.Hausman <dbl>
```
<sup>Created on 2020-09-04 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

In my view, the info contained in these diagnostics tests (here: Sargan test, etc.) are typically useful additions to a regression table. 

**Bottom line:** What do you think about including a "diagnostics = FALSE" (default) to `modelsummary()`, which could then get passed on the internal `glance()` call? (You'd probably also have to add some vars to the 'modelsummary::gof_map' data frame...)

**Comment by vincentarelbundock** [comment]
It's cool. I'm still in the honeymoon phase with `modelsummary`, and improving it still feels like a lot of fun.

I'm pretty resistant on principle to adding new arguments (one of the reasons I dislike some of the otherwise good competitors). But your goals are super easy to achieve by simply pushing `...` through to glance. With a 8 character commit, I think I can get you where you want to be.

Try this on master and let me know if it works for you:

```r
library(ivreg)
library(modelsummary)

demand = ivreg(Q ~ P + D | D + F + A, data = Kmenta)
msummary(demand, diagnostics=TRUE)
```

**Comment by vincentarelbundock** [comment]
And if you want to suggest sensible values for the default `gof_map`, I can make sure the diagnostics are printed nicely. See here for reference:

https://github.com/vincentarelbundock/modelsummary/blob/master/R/gof_map.R#L25

**Comment by grantmcdermott** [comment]
Ah, very nice. 

I was a nervous to suggest `...` at first because I know that the ellipsis is also getting passed to `tidy()` internally and didn't want misaligned arguments to cause errors, But on second thought that's never going to be a problem here because the broom tidiers do a good job of automatically ignoring superfluous args.

RE: sensible values for `gof_map`... Let me give this a bit of thought. In published tables you often see the tests listed according to what they do, rather than the names. E.g. "Overidentification test" instead of "Sargan". It might even be worth having them written in full:
- Weak instruments F-test
- Wu-Hausman endogeneity test
- Sargan overidentification test 

I'll ask some colleagues and get back to you.

PS. Congrats on the latest CRAN released! I'm delighted to see this package get the recognition it deserves.

**Comment by grantmcdermott** [comment]
One more thing on `gof_map`... I'd err on the side of just printing the test statistics, i.e. dropping the associated p-value rows. But do you think it would be possible to add stars to these statistics based on the (dropped) p-vals? So for the above example it might ultimately end up looking something like:

|                     |Model 1 |
|:--------------------|:-------|
|(coefs...)          | (vals...) |
|...          | ... |
| Sargan  overidentification test      | 2.983*  |
| Wu-Hausman endogenity test  | 11.422**   |

Or is that a huge PITA?

(If you're wondering about the missing weak instruments test, that's [an issue](https://github.com/tidymodels/broom/issues/922#issuecomment-687490778) on broom's side which I'm hoping to fix with a PR shortly.)

**Comment by vincentarelbundock** [comment]
I think the stars matching would be quite tricky to implement, because the current mechanism relies on having two matching vectors of estimates and p values. In your case, we'd need to mix and match based on the names of the different columns of the glance output. Obviously, that would completely model-specific, which is the kind of operation that is probably best left to broom.

Adding clean names to `gof_map` is trivial, so we should definitely do that, but if you want something more customized, this might be a good candidate for `glance_custom`. For instance:

```r
library(ivreg)
library(tibble)
library(modelsummary)

glance_custom.ivreg <- function(x) {
  stars <- c("***"=.01, "**"=.05, "*"=.1)
  gl <- glance(x, diagnostics=TRUE)
  s <- c(gl$statistic.Sargan, gl$statistic.Wu.Hausman)
  s <- modelsummary:::rounding(s, "%.3f")
  p <- c(gl$p.value.Sargan, gl$p.value.Wu.Hausman)
  out <- modelsummary:::make_stars(s, p, stars)
  out <- tibble(`Sargan` = out[1],
                `Wu Hausman` = out[2])
  out
}

mod = ivreg(Q ~ P + D | D + F + A, data = Kmenta)
modelsummary(mod)
```

Note that I exposed the `make_stars` function just now, so you'll need to reinstall from Github before running the code above.

**Comment by grantmcdermott** [comment]
Perfect.

Looking through a couple of AER (and similar level) pubs, I see a bit of variation in the way that this info is reported in regression tables. Some tables just list the test purpose (e.g. "Overidentification test"), others go by name (e.g. "Sargan *J*-stat"). 

If you are onboard, then maybe the best (safest?) thing to do from modelsummary's perspective, is to use both in `gof_map` by default. Users could then amend via their own custom gof mapping if so desired. My choices, including the common abbreviations that I've seen:

- "statistic.Weak.instruments" --> "Weak IV F-stat"
- "statistic.Wu.Hausman" --> "Endog. Wu-Hausman χ2-stat"
- "statistic.Sargan" --> "Overid. Sargan J-stat"

Again: Use, don't use. No offense taken either way and please feel free to close this if you're happy. Thanks again for indulging me ;-)

**Comment by vincentarelbundock** [comment]
Thanks, that's very useful! A couple things:

1. If I use a unicode χ, the html output works, but the table will no longer knit to PDF unless we specify `latex_engine:xelatex`. And even then, chi doesn't appear in the document.
2. What do you think of 3 decimals for p.values (omitted by default) and 1 for the tests (included by default)?
3. I imagine that most readers will  know what Wu-Hausman and Sargan J-stat are used for. Is "Endog." and "Overid." redundant? Would make the column less wide if we can shorten the string.

I was just bitten by excessive use of `...`. Here, it was because I push `...` through to `kableExtra::kbl` and it complains that `diagnostics` is an invalid argument.

Not sure what the cleanest way to do this, but I'm thinking that I should only pass "known" arguments forward.

**Comment by vincentarelbundock** [comment]
From Github Master, this looks pretty good to me:

``` r
library(ivreg)
library(modelsummary)

mod = ivreg(Q ~ P + D | D + F + A, data = Kmenta)
modelsummary(mod, "markdown", diagnostics=TRUE)
```

<table>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">Model 1</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(Intercept)</td>
<td style="text-align: left;">94.633</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(7.921)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">D</td>
<td style="text-align: left;">0.314</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0.047)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">P</td>
<td style="text-align: left;">-0.244</td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;">(0.096)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Num.Obs.</td>
<td style="text-align: left;">20</td>
</tr>
<tr class="even">
<td style="text-align: left;">R2</td>
<td style="text-align: left;">0.755</td>
</tr>
<tr class="odd">
<td style="text-align: left;">R2 Adj.</td>
<td style="text-align: left;">0.726</td>
</tr>
<tr class="even">
<td style="text-align: left;">Wu-Hausman Chi-Sq.</td>
<td style="text-align: left;">11.4</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sargan J-stat</td>
<td style="text-align: left;">3.0</td>
</tr>
</tbody>
</table>

**Comment by grantmcdermott** [comment]
LGTM!
